torch/csrc/jit/codegen/cuda/scheduler/mma_utils.cpp

```diff
@@ -0,0 +1,422 @@
+      // No support for swizzled inner dim for now.
+      //  Might need to add transpose swizzle here.
+      return false;
+    }
+    expr = leaf->definition();
+  }
+  return leaf == root;
+}
+
+} // namespace
+
+void checkDimSize(
+    TensorView* tv,
+    std::vector<int> axis,
+    std::vector<int> expect) {
+  ExpressionEvaluator const_eval(tv->fusion());
+  for (auto axis_index : c10::irange(axis.size())) {
+    auto id = tv->axis(axis[axis_index]);
+    auto maybe_extent = const_eval.evaluate(id->extent());
+    TORCH_CHECK(
+        maybe_extent.has_value(),
+        "Mma warp mapping: instruction tile has to be constant");
+    TORCH_CHECK(
+        maybe_extent.value() == expect[axis_index],
+        "Mma warp mapping: unexpected tile size at",
+        axis_index,
+        ":",
+        maybe_extent.value(),
+        "vs",
+        expect[axis_index]);
+  }
+}
+
+void WarpMmaSwizzler::scheduleMmaWarpOutput(
+    TensorView* tv,
+    MmaOptions options) {
+  auto macro = options.macro;
+  switch (macro) {
+    case MmaOptions::MacroType::Volta_16_16_4:
+      scheduleVoltaM16N16K4Fp32Output(tv, options);
+      if (tv->definition()->isA<MmaOp>()) {
+        setWarpMapped(tv, 5);
+      }
+      break;
+    default:
+      TORCH_CHECK(
+          false, "scheduleMmaWarp: unsupported mma option ", toString(macro));
+      break;
+  }
+}
+
+void WarpMmaSwizzler::scheduleOperandRead(TensorView* tv, MmaOptions options) {
+  // Schedules operand for inner most 3 contiguous dimensions
+  // Assumes M, N, K
+
+  switch (options.macro) {
+    case MmaOptions::MacroType::Volta_16_16_4:
+      scheduleVoltaOperandRead(tv, options);
+      break;
+    default:
+      TORCH_CHECK(false, "WarpMmaSwizzler: please specify macro");
+      break;
+  }
+}
+
+void WarpMmaSwizzler::setWarpMapped(TensorView* tv, int number_of_dims) {
+  for (int id : c10::irange(number_of_dims)) {
+    tv->axis(-id - 1)->toMmaSwizzled();
+  }
+}
+
+namespace {
+
+// Utility to check operand innermost scheduled dimensions
+void validateInnerMNK(TensorView* tv, MmaOptions options, int m, int n, int k) {
+  TORCH_INTERNAL_ASSERT(tv->nDims() >= 3);
+  TORCH_INTERNAL_ASSERT(canValidateIsInnerDim(
+      getMmaOperandRootDimension(tv, options, MmaDimension::M),
+      tv->axis(-3),
+      m));
+  TORCH_INTERNAL_ASSERT(canValidateIsInnerDim(
+      getMmaOperandRootDimension(tv, options, MmaDimension::N),
+      tv->axis(-2),
+      n));
+  TORCH_INTERNAL_ASSERT(canValidateIsInnerDim(
+      getMmaOperandRootDimension(tv, options, MmaDimension::K),
+      tv->axis(-1),
+      k));
+}
+
+void validateResultInnerMN(TensorView* tv, int m, int n) {
+  TORCH_INTERNAL_ASSERT(tv->nDims() >= 2);
+  int root_dim = tv->getMaybeRFactorDomain().size();
+  TORCH_INTERNAL_ASSERT(canValidateIsInnerDim(
+      tv->getMaybeRFactorDomain()[root_dim - 2], tv->axis(-2), m));
+  TORCH_INTERNAL_ASSERT(canValidateIsInnerDim(
+      tv->getMaybeRFactorDomain()[root_dim - 1], tv->axis(-1), m));
```

Comment:
Should this be `n`? Probably didn't show up since for volta `n == m`

Reply:
Good catch. Thanks for pointing this out. I actually fixed this on the Turing PR, and now propagated here as well.